[TIMOB-26038] Use JavaScriptCore framework API to implement proxies

sgtcoolguy opened the pull request:
**JIRA:** 
- https://jira.appcelerator.org/browse/TIMOB-26038
- https://jira.appcelerator.org/browse/TIMOB-23504 - iOS: Titanium proxies don't extend Object.prototype properly

**Description:**
This is a technical spike that extends, cleans up, and formalizes what I was looking at in #10069 

Basically, for some time now, Apple has had a supported mechanism for exposing native Obj-C classes/proxies to JavaScriptCore. It involves using protocols that extend from `JSExport` and then implementing those protocols in your classes. JavaScriptCore handles the grunt work of hooking up the methods/properties and converting types for you.

This PR begins migrating some of our proxies over using this mechanism and includes macros to simplify doing so, methods to handle dealing with conversions between *new* and *old* proxy types.

Some of the lessons learned:
- This simplifies the code quite a bit, **but** technically we could improve performance even more for constant properties by using the C API to assigns static values. (This and our old way basically use getters to wrap method calls to grab the property values).
- Ti.Buffer is special and allows indexed access to the contents. The new API has no means of doing this.
- When we expect to receive an "old-style" proxy, I had to typically hack the API a little to expect either `id` or `JSValue*` so that I can convert it. The JSC API doesn't know how to unwrap the old style proxies automatically itself. But as we move more proxies to the "new" style, we could do less of this and actually specify proper types.

sgtcoolguy committed:
refactor(ios): Move Ti.Accelerometer over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move Ti.Gesture over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move Ti.Geolocation over to new JavaScriptCore Obj-c based API for proxies
* Update some incorrect docs for the API
* Enable lastGeolocation property test

sgtcoolguy committed:
refactor(ios): Move Ti.Utils over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move Ti.Locale over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move top-level Ti object, Ti.API, console global over to JavaScriptCore Obj-c based API for proxies

Use a base ObjcProxy class for all new-style proxies
* Add helper methods for wrapping/unwrapping 'old style' proxies
* Add helper methods around events/listeners to base proxy class.
* Also add macros to make life simpler for declaring properties with deprecated getter/setter accessor methods

sgtcoolguy committed:
refactor(ios): Move Ti.Platform and Ti.Platform.DisplayCaps over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
fix(ios): Add missing imports to TiMediaSystemAlertProxy.m, TiUIActivityIndicatorStyleProxy.m, TiUIiOSLivePhoto.m

sgtcoolguy committed:
refactor(ios): Move Ti.Codec over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move Ti.Analytics over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy commented:
Note that I spent a lot of time/effort in making the commits/history super clean here. If we do merge this, it should be merged as "rebase and merge". Also, it should make it easier to review changes. The first commit lays all the ground work for hooking the new style proxies and uses them for a few core modules. The following commits are for converting other top-level modules over one by one.

build commented:
<!--
  1 failure:  Tests have failed...
  1 warning:  There is no linke...
  1 messages
  1 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  There is no linked JIRA ticket in the PR body. Please include the URL of the relevant JIRA ticket. If you need to, you may file a ticket on <a href="https://jira.appcelerator.org/secure/CreateIssue!default.jspa">JIRA</a>
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10381/1/artifact/dist/mobilesdk-8.0.0.v20181016133721-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| ios.Error | Native exception surfaced | 0.004 | expected '"Expected to throw exception" undefined "undefined"' to be 'Invalid type passed to function' |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

sgtcoolguy commented:
Interesting to note that the test for expecting a native error to be thrown when trying to set Ti.Geolocation.accuracy to null is failing because it no longer throws an Error! I'm curious how it's handling trying to convert null over under the hood (since the method/property is assigned the type `CLLocationAccuracy` which is a `double`).

sgtcoolguy committed:
refactor(ios): Move Ti.Blob over to new JavaScriptCore Obj-c based API for proxies
- Most of the Ti.Blob suite passes now (#toString() override doesn't work)
- Avoid Ti.Blob#toString(), use Ti.Blob.text in test suite

janvennemann commented:
@sgtcoolguy What's the plan here on how much you want to move over to JSC? All of it immediately? Just a bunch of our proxies at first that are easiest to convert? Will we be able to completely replace Kroll with this?

I also started some experimental work with a new Hyperloop version that tries to use JSC as much as possible and integrates our metabase to workaround the runtime restriction of JSC. Would be a perfect complement to this.

sgtcoolguy commented:
@janvennemann My initial goal was to explore this as a tech spike, and I think the earlier PR and Han's feedback sealed that this looked like a promising direction to go.

So for now, I'm iterating on this and attempting to get a good subset of the proxies moved over and still have the unit tests pass (and a nice clean commit history). I think once I've got a substantial set migrated (say, the modules already done above, plus Ti.Blob and Ti.Filesystem?) and the tests all pass, I'll put this PR up for merge review.

It's really up to us, @vijaysingh-axway and @mukherjee2 to decide on the sort of timeframe for this. My goal was to allow us to move this way "silently" (i.e. without impacting native modules or introducing any noticeable changes). If we can do so, then the first chunk might be able to land in 8.0.0, and we could continue to migrate more as we went. I think there's just so many APIs, it's not entirely feasible to move everything over at once.

janvennemann commented:
Partly moving over to this for 8.0.0 sounds good. Maybe we can even deprecate the "old" way of creating modules and proxies in favor of the new JSC way with 8.0.0.

sgtcoolguy committed:
refactor(ios): Move Ti.Filesystem module over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move Ti.Database module over to new JavaScriptCore Obj-c based API for proxies

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10381/8/artifact/dist/mobilesdk-8.0.0.v20181102081625-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| ios.Error | Native exception surfaced | 0.029 | expected '"Expected to throw exception" undefined "undefined"' to be 'Invalid type passed to function' |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

sgtcoolguy committed:
refactor(ios): Move rest of Ti.Database.* API over to new JavaScriptCore Obj-c based API for proxies

sgtcoolguy committed:
refactor(ios): Move Ti.Calendar.* APIs over to new JavaScriptCore Obj-c based API for proxies
* Update docs to fix incorrect info regarding properties that should be read-only, or incorrect types. Also update example code in docs.
* Remove methods from Ti.Calendar.Calendar that were deprecated in 7.0.0. This includes #getEventsInDate(), despite Android/docs not noting it as deprecated - iOS did report it as deprecated.
* Remove unused Ti.Calendar.Reminder type from iOS
* Return begin/end properties as Date objects on Ti.Calendar.Event, and Ti.Calendar.RecurrenceRule.end's endDate property. Docs/Android used Date, but iOS used String.

sgtcoolguy commented:
So, this PR is growing quite a bit - even if I'm trying to be good about structuring the commits. I think it'd be good to get view from @janvennemann and @vijaysingh-axway now and try and get this chunk in.

The only thing I worry about is the refactoring obviously changes the type hierarchy of these proxies. I think generally that should be ok as move native modules don't actually use any of these types, but TiBlob/TiFile might be used by them. So perhaps I should hold onto my TiBlob conversion here and we do that later? Or maybe, we say "oh well" and merge it in to 8.0.0 but don't bump the module api version because it's probably pretty unlikely many (or possibly any) native modules use TiBlob?

janvennemann commented on iphone/TitaniumKit/TitaniumKit/Sources/API/ObjcProxy.m:
```diff
@@ -0,0 +1,235 @@
+{
+  // remove all listeners JS side proxy
+  pthread_rwlock_wrlock(&m_listenerLock);
+  // FIXME: Need to unhoook managed values!
```
The docs state that if the `JSManagedValue` is not reachable anymore from the ObjC object graph it releases its underlying `JSValue`. So there should be no need to explicitly remove the object reference.

sgtcoolguy replied:
ok

janvennemann reviewed (commented):
Alright, i added a few notes, suggestions and questions. Man, this thing is a beast!

janvennemann reviewed (dismissed):
Added another round of feedback

sgtcoolguy commented on iphone/Classes/CodecModule.m:
```diff
@@ -30,26 +32,18 @@ - (NSString *)apiName
+  NSNumber *data = [dict[@"source"] toNumber];
+  NSString *type = [dict[@"type"] toString];
 
   int result = [TiUtils encodeNumber:data toBuffer:dest offset:position type:type endianness:byteOrder];
```
I assume we'll need to re-introduce some sanity check here, as I don't really not what would happen if there was no value. Presumably it'd error.

sgtcoolguy replied:
ok, I added back in some checking that required properties exist and validating types some.

sgtcoolguy commented on iphone/Classes/GeolocationModule.m:
```diff
@@ -458,95 +458,84 @@ - (BOOL)headingAvailable
+  // Need to use JSManagedValue here!
+  JSManagedValue *managedValue = [JSManagedValue managedValueWithValue:callback andOwner:self];
+  [callback.context.virtualMachine addManagedReference:managedValue withOwner:self];
+  [singleHeading addObject:managedValue];
```
It's entirely possible we don't need a managed value here...

sgtcoolguy commented on iphone/Classes/GestureModule.h:
```diff
@@ -1,26 +1,30 @@
+// methods
+// TODO: Add isFaceDown() and isFaceUp()?
+// - (BOOL)isFaceDown;
+// - (BOOL)isFaceUp;
```
I think these got removed a while back? I only see them listed for Windows now in our docs, and that may be only because we didn't remember to remove it there...

sgtcoolguy commented on iphone/Classes/PlatformModule.m:
```diff
@@ -105,10 +106,9 @@ - (void)dealloc
   RELEASE_TO_NIL(processorCount);
   RELEASE_TO_NIL(username);
-  RELEASE_TO_NIL(address);
+  //  RELEASE_TO_NIL(address);
```
There's no variable/field. It's declared as a property, but is implemented as getter/setter.

sgtcoolguy commented on iphone/Classes/CalendarModule.m:
```diff
@@ -214,12 +210,10 @@ - (void)requestAuthorization:(id)args forEntityType:(EKEntityType)entityType
+    [callback callWithArguments:@[ propertiesDict ]];
     return;
   }
+  // FIXME: I think we need to store the callback as a JSManagedValue!
```
ok

sgtcoolguy commented on iphone/Classes/TiDatabaseResultSetProxy.m:
```diff
@@ -9,35 +9,20 @@
-    [results close];
-  }
-  RELEASE_TO_NIL(results);
+  [self close];
```
`close:` will release the database

sgtcoolguy commented on iphone/Classes/TiCalendarEvent.m:
```diff
@@ -43,176 +45,200 @@ - (NSString *)apiName
+    }                                                          \
+                                                               \
+    ENSURE_EVENT                                               \
+    if (currEvent == NULL) {                                   \
```
it should throw an exception yes. so presumably the NULL path wouldn't get followed unless there's some handler that allow it to resume.

sgtcoolguy commented on iphone/TitaniumKit/TitaniumKit/Sources/API/ObjcProxy.m:
```diff
@@ -0,0 +1,235 @@
+    if (listenersForType == nil) {
+      return;
+    }
+    // FIXME: looks like we need to handle bubble logic/etc. See other fireEvent impl
```
so for now I punted on this because none of the proxies I converted made use of the bubbling stuff. That's effectively a UI thing, as far as I can tell.

There's a lot of ugly C-API JSC code around TiBindingEvent.m to do the old event firing, with lots of various fireEvent methods in TiProxy too.

But from what I can tell, the bubbles/propagate stuff was really meant to bubble events up the view hierarchy by using `parentForBubbling` which I only see overrides in Ti.Ui.View and Ti.UI.TableViewRow.

So, I think for now we can continue to punt on this bubbling stuff - **but it's entirely possible I did miss some other code around setting default event values when not specified.**

janvennemann replied:
Yeah, that totally makes sense. Apart from UI there shouldn't be any need for event bubbling at all. I guess then it's fine to but this aside until we touch the UI stuff.

sgtcoolguy commented on iphone/TitaniumKit/TitaniumKit/Sources/API/KrollBridge.m:
```diff
@@ -513,63 +367,148 @@ - (void)willStartNewContext:(KrollContext *)kroll
+    //        [ti noteKrollObject:ko forKey:key];
+    //        [ti setStaticValue:ko forKey:key purgable:NO];
+    //      }
+    //    }
```
Is there any way for me to be able to test this? I can try and uncomment and get it working, but I'd like to be able to actually try it out to make sure it works...

build commented:
<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@8.0.0 test /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10381
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10381/tests/Resources/ti.geolocation.test.js
  166:9  warning  'returnValue' is assigned a value but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)


Running "checkFormat:ios" (checkFormat) task
Fatal error: Error: Formatting incorrect on "iphone/Classes/TiNetworkHTTPClientProxy.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='265' length='251'>#import "TiNetworkHTTPClientProxy.h"&#10;#import "NetworkModule.h"&#10;#import &lt;TitaniumKit/Mimetypes.h>&#10;#import &lt;TitaniumKit/TiApp.h>&#10;#import &lt;TitaniumKit/TiBase.h>&#10;#import &lt;TitaniumKit/TiBlob.h>&#10;#import &lt;TitaniumKit/TiFile.h>&#10;#import &lt;TitaniumKit/TiUtils.h></replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

build commented on tests/Resources/ti.geolocation.test.js:
```diff
@@ -0,0 +1,317 @@
+
+	it('#getLastGeolocation()', function () {
+		should(Ti.Geolocation).have.a.property('getLastGeolocation').which.is.a.Function;
+		const returnValue = Ti.Geolocation.getLastGeolocation();
```
<!--
  0 failure: 
  1 warning:  tests/Resources/t...
  
  
  DangerID: danger-id-default;
  File: tests/Resources/ti.geolocation.test.js;
  Line: 166;
-->

- :warning: tests/Resources/ti.geolocation.test.js line 166 – 'returnValue' is assigned a value but never used. (no-unused-vars)

sgtcoolguy commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TopTiModule.m:
```diff
@@ -85,7 +92,8 @@ - (TiBuffer *)createBuffer:(id)arg
       case BAD_SRC_OFFSET: { // String length == 0 : return our empty buffer (no string encoded into it)
-        return buffer;
+        KrollObject *o = [[[KrollObject alloc] initWithTarget:buffer context:context] autorelease];
+        return [JSValue valueWithJSValueRef:JSObjectMake(contextRef, KrollObjectClassRef, o) inContext:objcJsContext];
```
ok, I have it now creating a KrollObject and then returning the `[krollObject jsobject]` so we're not double-generating js objects here.

sgtcoolguy committed:
feat(ios): Add macros for validating properties and types of properties/defaults on JSValue* object arguments to methods

sgtcoolguy committed:
fix(ios): Add validation of required properties on passed in objects for Ti.Codec methods.

sgtcoolguy committed:
fix(ios): Fix how we 'throw' exceptions under Obj-C proxies to set the error on the JSContext* exception property and return nil immediately.

sgtcoolguy committed:
fix(ios): collect native stacktrace when using new obj-c proxy exception macros

janvennemann commented:
@sgtcoolguy @mukherjee2 This looks good to go now! However since this is a massive change in our core i think an RC period of ~1 month would make sense for this and not just the usual one or two weeks. Is this doable in the 8.0.0 timeframe? It does not contain breaking changes so we could push it out to 8.1.0, but since this is such a huge change it would be good to get this in as soon as possible (see the first merge conflicts that are popping up)

sgtcoolguy committed:
Merge branch 'master' into ios-objc-api

sgtcoolguy committed:
Merge branch 'master' into ios-objc-api

sgtcoolguy committed:
Merge branch 'master' into ios-objc-api
